utils/atomic-data-publisher/src/service.rs

```diff
@@ -0,0 +1,513 @@
+    ).await?);
+
+    let (shutdown_sender, shutdown_signal) = tokio::sync::watch::channel(false);
+
+    Ok(Self {
+      database,
+      publisher,
+      metrics,
+      config,
+      shutdown_signal,
+      shutdown_sender,
+    })
+  }
+
+  pub async fn run(&self) -> Result<(), AtomicDataError> {
+    self.health_check().await?;
+    let (handles, _metrics_bind_addr) = self.spawn_background_tasks().await?;
+    let shutdown_sender = self.shutdown_sender.clone();
+    let signal_handle = tokio::spawn(async move {
+      tokio::select! {
+        _ = signal::ctrl_c() => {
+          info!("Received Ctrl+C within service, initiating graceful shutdown");
+        }
+        _ = async {
+          signal::unix::signal(signal::unix::SignalKind::terminate())
+            .expect("failed to install signal handler")
+            .recv()
+            .await
+        } => {
+          info!("Received SIGTERM within service, initiating graceful shutdown");
+        }
+      }
+
+      if let Err(e) = shutdown_sender.send(true) {
+        error!("Failed to send internal shutdown signal: {}", e);
+      }
+    });
+
+    let mut shutdown_signal = self.shutdown_signal.clone();
+    tokio::select! {
+        _ = shutdown_signal.changed() => {
+            info!("Shutdown signal received");
+        }
+        result = futures::future::try_join_all(handles) => {
+            match result {
+                Ok(results) => {
+                    let failed_tasks = results.iter().filter(|r| r.is_err()).count();
+                    if failed_tasks > 0 {
+                        warn!("{} out of {} background tasks failed", failed_tasks, results.len());
+                    } else {
+                        info!("All background tasks completed successfully");
+                    }
+                }
+                Err(e) => error!("Background task failure: {}", e),
+            }
+        }
+        _ = signal_handle => {
+          // Signal handler completed
+        }
+    }
+
+    self.perform_graceful_shutdown().await?;
+    info!("Atomic Data Publisher service stopped");
+    Ok(())
+  }
+
+  async fn spawn_background_tasks(&self) -> Result<(Vec<tokio::task::JoinHandle<Result<(), AtomicDataError>>>, String), AtomicDataError> {
+    let mut handles = Vec::new();
+    let metrics_bind_addr = format!("0.0.0.0:{}", self.config.service.port);
+    let metrics_handle = {
+      let metrics = self.metrics.clone();
+      let shutdown_signal = self.shutdown_signal.clone();
+      let bind_addr = metrics_bind_addr.clone();
+      tokio::spawn(async move {
+        if let Err(e) = metrics.serve_metrics(&bind_addr, shutdown_signal).await {
+          error!("Metrics server error: {}", e);
+          return Err(AtomicDataError::NetworkError(e.to_string()));
+        }
+        Ok(())
+      })
+    };
+    handles.push(metrics_handle);
+    info!("Metrics server started on {}", metrics_bind_addr);
+
+    let polling_handle = {
+      let service = self.clone();
```

Comment:
Feels a bit weird for the service to clone itself to start a polling loop. Should the background tasks be another trait/service?